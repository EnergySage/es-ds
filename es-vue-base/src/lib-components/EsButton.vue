<template>
    <b-button
        :class="{ 'inline': inline }"
        :variant="computedVariant"
        :size="size"
        v-bind="$attrs"
        v-on="$listeners">
        <!--
            @slot Button Content
            @binding {string} text or html of the button content
        -->
        <slot />
    </b-button>
</template>

<script lang="js">
import { BButton } from 'bootstrap-vue';

export default {
    name: 'EsButton',
    components: {
        BButton,
    },
    props: {
        /**
         * Button Variant
         *
         */
        variant: {
            type: String,
            default: 'primary',
            required: false,
        },
        /**
         * Outline only button
         */
        outline: {
            type: Boolean,
            default: false,
        },
        /**
         * Size of button
         */
        size: {
            type: String,
            validator: (val) => ['lg', 'md', 'sm'].includes(val),
            default: 'md',
        },
        /**
         * Only works for 'link' variant buttons.
         * Removes the fixed height and padding so the button can fit nicely within a block of text.
         */
        inline: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        /**
         * Compute the Variant
         * used for when you want to `outline` something but don't know the
         * theme's default variant
         *
         * @returns { string } computed variant based on outline prop
         */
        computedVariant() {
            return `${this.outline && this.variant !== 'link' ? 'outline-' : ''}${this.variant}`;
        },
    },
};
</script>

<style lang="scss" scoped>
.btn-link.inline {
    /* use normal CSS here so users can override with utility classes as necessary */
    border: 0;
    height: auto;
    line-height: inherit;
    margin: 0;
    padding: 0;
    vertical-align: baseline;
}
</style>
