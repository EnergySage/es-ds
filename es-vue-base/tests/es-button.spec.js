import { mount } from '@vue/test-utils';
import EsButton from '../src/lib-components/es-button.vue';
import jestVue from './jest.vue.config';


describe('EsButton', () => {
    const variantOptions = ['primary', 'secondary', 'danger', 'warning', 'info', 'light', 'dark', 'link'];
    const outlineOptions = [true, false];
    const slots = {
        default: 'Main Content',
    };
    const paramaters = [];
    variantOptions.forEach((variant) => {
        outlineOptions.forEach((outline) => {
            paramaters.push({
                variant,
                outline,
            });
        });
    });

    test.each(paramaters)(
        '<es-button variant="$variant" outline="$outline" />',
        async ({
            variant, outline,
        }) => {
            const wrapper = mount(EsButton, {
                ...jestVue,
                slots,
                propsData: {
                    variant,
                    outline,
                },
            });
            const a11y = await axe(wrapper.element);

            expect(wrapper.vm).toBeTruthy();
            expect(wrapper.html()).toMatchSnapshot();
            expect(a11y).toHaveNoViolations();
        },
    );

    test('<EsButton uppercase="true" />', () => {
        const wrapper = mount(EsButton, {
            ...jestVue,
            slots,
            propsData: {
                outline: true,
            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('<EsButton size="sm" />', () => {
        const wrapper = mount(EsButton, {
            ...jestVue,
            slots,
            propsData: {
                size: 'sm',
            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
