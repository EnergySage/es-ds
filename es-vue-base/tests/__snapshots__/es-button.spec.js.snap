// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsButton <EsButton size="sm" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-primary btn-sm  text-uppercase text-nowrap\\" size=\\"sm\\">Main Content</button>"`;

exports[`EsButton <EsButton uppercase="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-primary  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="danger" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-danger  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="danger" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-danger  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="dark" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-dark  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="dark" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-dark  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="info" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-info  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="info" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-info  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="light" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-light  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="light" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-light  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="link" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-link  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="link" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-link  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="primary" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-primary  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="primary" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-primary  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="secondary" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-secondary  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="secondary" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-secondary  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="warning" outline="false" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-warning  text-uppercase text-nowrap\\">Main Content</button>"`;

exports[`EsButton <es-button variant="warning" outline="true" /> 1`] = `"<button type=\\"button\\" class=\\"btn btn-outline-warning  text-uppercase text-nowrap\\">Main Content</button>"`;
